text/0000-alloc-me-maybe.md

```diff
@@ -0,0 +1,547 @@
+/// that it managed to insert until the failure. The result will contain
+/// the iterator, having been run up until the failure point. If the iterator
+/// has been run at all, the last element yielded will also be returned.
+fn extend<I: IntoIter<Item=T>>(&mut self, iter: I)
+    -> Result<(), (I::IntoIter, Option<T>, Err)>
+```
+
+Note that this API only even works because Iterator's signature currently guarantees that the yielded elements outlive the iterator. This would not be the case if we ever moved to support so-called "streaming iterators", which yield elements that point into themselves.
+
+
+
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Due to the diversity of requirements between our user profiles, there isn't any one-size fits all solution. This RFC proposes two solutions which will require minimal work for maximal impact:
+
+* For the server users, an `oom=panic` configuration, in the same vein as the `panic=abort`.
+* For everyone else, add `try_reserve` and `try_reserve_exact` as standard collection APIs.
+
+
+
+## oom=panic
+
+Applying this configuration in a Cargo.toml would change the behaviour of the global allocator's `oom()` function, which currently aborts, to instead panic. As discussed in the Server user profile, this would allow OOM to be handled at task boundaries with minimal effort for server developers, and no effort from library maintainers.
+
+If using a thread-per-task model, OOMs will be naturally caught at the thread boundary. If using a different model, tasks can be isolated using the `thread::catch_unwind` or `Future::catch_unwind` APIs.
+
+We expose a flag, rather than changing the default, because we maintain that *by default* Rust programmers should not be trying to recover from allocation failures.
+
+For instance, a project which desires to work this way would add this to their Cargo.toml:
+
+```toml
+[profile]
+oom = panic
+```
+
+And then in their application, do something like this:
+
+```rust
+fn main() {
+    set_up_event_queue();
+    loop {
+        let event = get_next_event();
+        let result = ::std::panic::catch_unwind(|| {
+            process_event(&mut event)
+        });
+
+        if let Err(err) = result {
+            if let Some(message) = err.downcast_ref::<&str>() {
+                eprintln!("Task crashed: {}", message);
+            } else if let Some(message) = err.downcast_ref::<String>() {
+                eprintln!("Task crashed: {}", message);
+            } else {
+                eprintln!("Task crashed (unknown cause)");
+            }
+
+            // Handle failure...
+        }
+    }
+}
+```
+
+
+
+
+## try_reserve
+
+`try_reserve` and `try_reserve_exact` would be added to `HashMap`, `Vec`, `String`, and `VecDeque`. These would have the exact same APIs as their infallible counterparts, except that OOM would be exposed as an error case, rather than a call to `Alloc::oom()`. They would have the following signatures:
```

Comment:
Clarification request: You'll only be adding `fn try_reserve_exact` to the types that already supply `fn reserve_exact`, right?